custom_components/waste_collection_schedule/waste_collection_schedule/source/lipizzanerheimat_at.py

```diff
@@ -0,0 +1,335 @@
+
+    def api_hash(self, endpoint, payload):
+        return self.sha1(f"{APISECRET}@{endpoint}@{payload}")
+
+    @staticmethod
+    def get_basic_auth_header(user, password):
+        credentials = f"{user}:{password}"
+        base64_credentials = base64.b64encode(
+            credentials.encode("utf-8")).decode("utf-8")
+        return f"Basic {base64_credentials}"
+
+    def sha1(self, input_str):
+        return hashlib.sha1(input_str.encode('utf-8')).hexdigest()
+
+    def build_header(self, url, payload):
+        apihash = self.api_hash(url, payload)
+        headers = self.default_headers
+        headers['apihash'] = apihash
+        return headers
+
+    def fetch_communities(self):
+
+        url = 'api/mobile-app/getcommunities'
+        payload = 'language=de'
+
+        headers = self.build_header(url, payload)
+
+        communities = requests.request(
+            "POST", f'{API_URL}/{url}', headers=headers, data=payload).json()['data']
+
+        return next((c['community_id'] for c in communities if c['community_title'] == self.town), None)
+
+    def fetch_calendar_id_by_name(self):
+
+        url = 'api/mobile-app/garbagecalendarindexwithstreets'
+        payload = f'language=de&communities=234|110|110|111|112|241|238|119|113|236|114|237|115|116|117|235&garbage_calendar_id=1'
+
+        headers = self.build_header(url, payload)
+
+        calendarData = requests.request(
+            "POST", f'{API_URL}/{url}', headers=headers, data=payload).json()['data']
+
+        return next((c['garbage_calendar_id']
+                     for c in calendarData['garbage_calendars'] if c['garbage_calendar_title'] == self.map_name), None)
+
+    def fetch_calendar_id(self):
+
+        community_id = self.fetch_communities()
+
+        if (community_id == None):
+            raise SourceArgumentNotFound(
+                [], "Town/Street combination not found.")
+
+        url = 'api/mobile-app/garbagecalendarindexwithstreets'
+        payload = f'language=de&communities={community_id}&garbage_calendar_id=1'
+
+        headers = self.build_header(url, payload)
+
+        response = requests.request(
+            "POST", f'{API_URL}/{url}', headers=headers, data=payload)
+
+        calendarData = response.json()['data']
+
+        return next((c['garbage_calendars_available'][0]
+                     for c in calendarData['street_structure'] if c['garbage_calendar_street'] == self.street), None)
+
+    def _first_setup(self):
+
+        r = requests.get(API_URL)
+
+        if (self.street and self.town):
+            garbage_calendar_id = self.fetch_calendar_id()
+        elif (self.map_name):
+            garbage_calendar_id = self.fetch_calendar_id_by_name()
+        elif (self.garbage_calendar_id):
+            garbage_calendar_id = self.garbage_calendar_id
+        else:
+            raise SourceArgumentException(
+                [], "Please provide either `garbage_calendar_id`, `street` and `town`, or `map_name`.")
```

Comment:
The first argument needs to be the string of the parameter as used in the __init__

This should probably get a type hint and a docstring


same for SourceArgumentNotFound but the first argument is the string of the variable name the second one the value and the third argument is an optional message